uml4net.xmireader/Readers/XmiReader.cs

```diff
@@ -0,0 +1,234 @@
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+// </copyright>
+// ------------------------------------------------------------------------------------------------
+
+namespace uml4net.xmi.Readers
+{
+    using Cache;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Xml;
+
+    using Microsoft.Extensions.Logging;
+    using System.Threading.Tasks;
+    using uml4net.POCO.Packages;
+
+    using uml4net;
+    using xmi;
+
+    /// <summary>
+    /// The purpose of the <see cref="XmiReader"/> is to provide a means to read (deserialize)
+    /// an UML 2.5.1 model from XMI
+    /// </summary>
+    public class XmiReader : IXmiReader
+    {
+        /// <summary>
+        /// The <see cref="IAssembler"/>
+        /// </summary>
+        private readonly IAssembler assembler;
+
+        /// <summary>
+        /// The <see cref="ILogger"/> used to log
+        /// </summary>
+        private readonly ILogger<XmiReader> logger;
+
+        /// <summary>
+        /// The <see cref="IXmiElementReader{T}"/>of <see cref="IPackage"/>
+        /// </summary>
+        private readonly IXmiElementReader<IPackage> packageReader;
+
+        /// <summary>
+        /// The <see cref="IXmiElementReader{T}"/>of <see cref="IModel"/>
+        /// </summary>
+        private readonly IXmiElementReader<IModel> modelReader;
+
+        /// <summary>
+        /// The <see cref="IExternalReferenceResolver"/>
+        /// </summary>
+        private readonly IExternalReferenceResolver externalReferenceResolver;
+
+        /// <summary>
+        /// The <see cref="IXmiReaderScope"/>
+        /// </summary>
+        private readonly IXmiReaderScope scope;
+
+        /// <summary>
+        /// The <see cref="IXmiReaderCache"/>
+        /// </summary>
+        private readonly IXmiReaderCache cache;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmiReader"/> class.
+        /// </summary>
+        /// <param name="assembler">The <see cref="IAssembler"/></param>
+        /// <param name="cache">The <see cref="IXmiReaderCache"/></param>
+        /// <param name="logger">
+        /// The (injected) <see cref="ILogger{T}"/> used to setup logging
+        /// </param>
+        /// <param name="packageReader">The <see cref="IXmiElementReader{T}"/>of <see cref="IPackage"/></param>
+        /// <param name="modelReader">The <see cref="IXmiElementReader{T}"/>of <see cref="IModel"/></param>
+        /// <param name="externalReferenceResolver">The <see cref="IExternalReferenceResolver"/></param>
+        /// <param name="scope">The <see cref="IXmiReaderScope"/></param>
+        public XmiReader(IAssembler assembler, IXmiReaderCache cache, ILogger<XmiReader> logger, IXmiElementReader<IPackage> packageReader,
```

Comment:
i think we are going to run into trouble here. When using constructor injection we run the risk that the list of arguments will become very long. I think even that a UML model can start with any valid XML Element, meaning we would have to inject all readers here.

Perhaps property injection, using a DI container like Autofac, is a better approach that will make it more readable? https://autofac.readthedocs.io/en/latest/register/prop-method-injection.html#property-injection